components/usage/pkg/apiv1/usage.go

```diff
@@ -118,6 +118,105 @@ func (s *UsageService) ListBilledUsage(ctx context.Context, in *v1.ListBilledUsa
 	}, nil
 }
 
+func (s *UsageService) ListUsage(ctx context.Context, in *v1.ListUsageRequest) (*v1.ListUsageResponse, error) {
+	to := time.Now()
+	if in.To != nil {
+		to = in.To.AsTime()
+	}
+	from := to.Add(-maxQuerySize)
+	if in.From != nil {
+		from = in.From.AsTime()
+	}
+
+	if from.After(to) {
+		return nil, status.Errorf(codes.InvalidArgument, "Specified From timestamp is after To. Please ensure From is always before To")
+	}
+
+	if to.Sub(from) > maxQuerySize {
+		return nil, status.Errorf(codes.InvalidArgument, "Maximum range exceeded. Range specified can be at most %s", maxQuerySize.String())
+	}
+
+	var order db.Order = db.DescendingOrder
+	if in.Order == v1.ListUsageRequest_ORDERING_ASCENDING {
+		order = db.AscendingOrder
+	}
+
+	var limit int64 = 1000
+	var page int64 = 0
+	var offset int64 = 0
+	if in.Pagination != nil {
+		limit = in.Pagination.PerPage
+		page = in.Pagination.Page
+		offset = limit * (int64(math.Max(0, float64(page-1))))
+	}
+
+	listUsageResult, err := db.FindUsage(ctx, s.conn, &db.FindUsageParams{
+		AttributionId: db.AttributionID(in.GetAttributionId()),
+		From:          from,
+		To:            to,
+		Order:         order,
+		Offset:        offset,
+		Limit:         limit,
+	})
+	logger := log.Log.
+		WithField("attribution_id", in.AttributionId).
+		WithField("perPage", limit).
+		WithField("page", page).
+		WithField("from", from).
+		WithField("to", to)
+	if err != nil {
+		logger.WithError(err).Error("Failed to fetch usage.")
+		return nil, status.Error(codes.Internal, "unable to retrieve usage")
+	}
+
+	var usageData []*v1.Usage
+	for _, usageRecord := range listUsageResult {
+		kind := v1.Usage_KIND_WORKSPACE_INSTANCE
+		if usageRecord.Kind == db.InvoiceUsageKind {
+			kind = v1.Usage_KIND_INVOICE
+		}
+		usageDataEntry := &v1.Usage{
+			Id:            usageRecord.ID.String(),
+			AttributionId: string(usageRecord.AttributionID),
+			EffectiveTime: timestamppb.New(usageRecord.EffectiveTime.Time()),
+			// convert cents back to full credits
+			Credits:             float64(usageRecord.CreditCents) / 100,
+			Kind:                kind,
+			WorkspaceInstanceId: usageRecord.WorkspaceInstanceID.String(),
+			Draft:               usageRecord.Draft,
+			Metadata:            string(usageRecord.Metadata),
+		}
+		usageData = append(usageData, usageDataEntry)
+	}
+
+	usageSummary, err := db.GetUsageSummary(ctx, s.conn,
+		db.AttributionID(in.GetAttributionId()),
```

Comment:
I think we should first attempt to parse the attribution using `db.ParseAttributionID` which does a sanity check that the ID is valid.